editor/src/clj/editor/code/view.clj

```diff
@@ -2641,6 +2684,129 @@
       (ui/timer-stop! timer)
       (reset! state nil))))
 
+(defn- consume-breakpoint-popup-events [^Event e]
+  (when-not (and (instance? KeyEvent e)
+                 (= KeyEvent/KEY_PRESSED (.getEventType e))
+                 (= KeyCode/ESCAPE (.getCode ^KeyEvent e)))
+    (.consume e)))
+
+(defn- breakpoint-editor-view [^Canvas canvas {:keys [edited-breakpoint gutter-metrics layout]}]
+  (let [[^double gutter-width ^double gutter-margin] gutter-metrics
+        spacing 4
+        padding 8
+        anchor (.localToScreen canvas
+                               (- gutter-width
+                                  (* gutter-margin 0.5)
+                                  12) ;; shadow offset
+                               (+ (* ^double (data/line-height (:glyph layout)) 0.5)
+                                  (data/row->y layout (data/breakpoint-row edited-breakpoint))))]
+    {:fx/type fxui/with-popup
+     :desc {:fx/type fxui/ext-value :value (.getWindow (.getScene canvas))}
+     :showing true
+     :anchor-x (.getX anchor)
+     :anchor-y (.getY anchor)
+     :anchor-location :window-top-left
+     :auto-hide true
+     :auto-fix true
+     :hide-on-escape true
+     :on-auto-hide {:event :cancel}
+     :consume-auto-hiding-events true
+     :event-handler consume-breakpoint-popup-events
+     :content
+     [{:fx/type fx.v-box/lifecycle
+       :stylesheets [(str (io/resource "editor.css"))]
+       :children
+       [{:fx/type fx.v-box/lifecycle
+         :style-class "breakpoint-editor"
+         :fill-width false
+         :spacing -1
+         :children
+         [{:fx/type fx.region/lifecycle
+           :view-order -1
+           :style-class "breakpoint-editor-arrow"
+           :min-width 10
+           :min-height 10}
+          {:fx/type fx.stack-pane/lifecycle
+           :children
+           [{:fx/type fx.region/lifecycle
+             :style-class "breakpoint-editor-background"}
+            {:fx/type fx.v-box/lifecycle
+             :style-class "breakpoint-editor-content"
+             :spacing padding
+             :children
+             [{:fx/type fx.label/lifecycle
+               :style-class ["label" "breakpoint-editor-label" "breakpoint-editor-header"]
+               :text (format "Breakpoint on line %d" (inc ^long (data/breakpoint-row edited-breakpoint)))}
```

Comment:
`data/breakpoint-row` just `asserts` that the region is of the expected type before returning the `row`, but since we know this will be a breakpoint region we can use `data/CursorRange->line-number` instead.

```suggestion
               :text (format "Breakpoint on line %d" (data/CursorRange->line-number edited-breakpoint))}
```